unicon_backend/evaluator/tasks/programming.py

```diff
@@ -80,20 +91,50 @@ def run(self, input_files: list[File], environment: ProgrammingEnvironment):
 
         print(f"Testcase File (run.py):\n {file}")
 
+        user_result = self._run_code(file, environment)
+
+        check_steps = [step for step in self.steps if isinstance(step, CheckOutputStep)]
+        for check_step in check_steps:
+            if isinstance(check_step.value, OutputFile):
+                if check_step.value.value != user_result["stdout"]:
+                    user_result["status"] = "WA"
+                    break
+            if isinstance(check_step.value, File):
+                sample_file = template.render(
+                    prepend="", artifacts=[check_step.value], run_funcs=run_funcs
+                )
+                correct_output = self._run_code(sample_file, environment)
+                if correct_output["stdout"] != user_result["stdout"]:
+                    user_result["status"] = "WA"
+                    break
+
+        return user_result
+
+    def _run_code(self, file: str, environment: ProgrammingEnvironment):
         request = Request(
             files=[File(file_name="run.py", content=file)],
             environment=environment,
             entrypoint="run.py",
         )
 
-        if RUNNER_URL:
-            resp = requests.post(
-                f"{RUNNER_URL}/submissions", data=request.model_dump_json()
-            )
-            print(resp.json())
-        else:
-            # TEMP
+        if not RUNNER_URL:
             print("WARN: No programming task runner set!")
+            return
+
+        # 1. Run the user's code.
+        resp = requests.post(
+            f"{RUNNER_URL}/submissions", data=request.model_dump_json()
+        )
+        submission_id = resp.json()["submission_id"]
+        while True:
```

Comment:
instead of active polling we could just return the state of the test case as `PENDING`? I can add a way in backend to only rerun/poll a specific test case or task